metadata-ingestion/src/datahub/ingestion/source/sagemaker.py

```diff
@@ -0,0 +1,300 @@
+        List all feature groups in SageMaker.
+        """
+
+        feature_groups = []
+
+        # see https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/sagemaker.html#SageMaker.Client.list_feature_groups
+        paginator = self.sagemaker_client.get_paginator("list_feature_groups")
+        for page in paginator.paginate():
+            feature_groups += page["FeatureGroupSummaries"]
+
+        return feature_groups
+
+    def get_feature_group_details(self, feature_group_name: str) -> Dict[str, Any]:
+        """
+        Get details of a feature group (including list of component features).
+        """
+
+        # see https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/sagemaker.html#SageMaker.Client.describe_feature_group
+        feature_group = self.sagemaker_client.describe_feature_group(
+            FeatureGroupName=feature_group_name
+        )
+
+        # use falsy fallback since AWS stubs require this to be a string in tests
+        next_token = feature_group.get("NextToken", "")
+
+        # paginate over feature group features
+        while next_token:
+            next_features = self.sagemaker_client.describe_feature_group(
+                FeatureGroupName=feature_group_name, NextToken=next_token
+            )
+            feature_group["FeatureDefinitions"].append(
+                next_features["FeatureDefinitions"]
+            )
+            next_token = feature_group.get("NextToken", "")
+
+        return feature_group
+
+    def get_feature_group_wu(
+        self, feature_group_details: Dict[str, Any]
+    ) -> MetadataWorkUnit:
+        """
+        Generate an MLFeatureTable workunit for a SageMaker feature group.
+
+        Parameters
+        ----------
+            feature_group_details:
+                ingested SageMaker feature group from get_feature_group_details()
+        """
+
+        feature_group_name = feature_group_details["FeatureGroupName"]
+
+        feature_group_snapshot = MLFeatureTableSnapshot(
+            urn=builder.make_ml_feature_table_urn("sagemaker", feature_group_name),
+            aspects=[],
+        )
+
+        feature_group_snapshot.aspects.append(
+            MLFeatureTablePropertiesClass(
+                description=feature_group_details.get("Description"),
+                # non-primary key features
+                mlFeatures=[
+                    builder.make_ml_feature_urn(
+                        feature_group_name,
+                        feature["FeatureName"],
+                    )
+                    for feature in feature_group_details["FeatureDefinitions"]
+                    if feature["FeatureName"]
+                    != feature_group_details["RecordIdentifierFeatureName"]
+                ],
+                mlPrimaryKeys=[
+                    builder.make_ml_primary_key_urn(
+                        feature_group_name,
+                        feature_group_details["RecordIdentifierFeatureName"],
```

Comment:
Is this always one feature in SageMaker?

Reply:
Yep – they make you specify one when a feature group is created